KEP-5229: Asynchronous API calls during scheduling

macsko opened the pull request:
<!-- 
	Please use the following format when naming your PR
	< Issue Number >:< Issue Description >
	e.g. KEP-000: adding beta graduation criteria
	
	Avoid using phrases like `fixes #NNNN` in the description
	unless the pull request is to change the KEP status to 
	implemented or KEP has been deprecated.
-->

<!-- short description of work done in PR e.g. updating milestone, adding new KEP, adding test requirements… -->  
- One-line PR description: Add KEP-5229

<!-- link to the k/enhancements issue -->
- Issue link: https://github.com/kubernetes/enhancements/issues/5229

<!-- other comments or additional information -->
- Other comments:

macsko commented:
/cc @dom4ha @sanposhiho 

I have written some possible approaches to these API calls to start the discussion. I will not be visibly active for the next three weeks, but feel free to comment.

macsko commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,961 @@
+- `nominatedNodeName` scenario support would require more effort in (1.1) or (1.2).
+
+
+#### 2.2: Make the API calls queued
```
I feel like queueing API calls might be a good direction long term and probably won't be that hard to implement.

sanposhiho reviewed (commented):
Great summary! I'll take time reading through it and put some comments after that.

wojtek-t commented on keps/prod-readiness/sig-scheduling/5229.yaml:
```diff
@@ -0,0 +1,3 @@
+kep-number: 5229
+alpha:
+  approver: ""
```
Feel free to add me here

wojtek-t commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,961 @@
+This would require implementing advanced logic for queueing API calls in the kube-scheduler and migrating **all** pod-based API calls done during scheduling to this method,
+potentially including the binding API call. The new component should be able to resolve any conflicts in the incoming API calls as well as parallelize them properly,
+e.g., don't parallelize two updates of the same pod. This requires [making the API calls queued](#22-make-the-api-calls-queued) or
+[sending API calls through a kube-scheduler's cache](#23-send-api-calls-through-a-kube-schedulers-cache) to be implemented.
```
Conceptually what you need here is something pretty similar to DeltaFIFO we use in client-go:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/delta_fifo.go

We wouldn't operate on object, but on their updates, but that's conceptually exactly what you need:
- tracking changes to a given object all together
- a concept of deduping (if I already scheduled a pod but didn't yet sent the failures, don't send them; if I have multiple failures not yet sent to report, is the last one only valid?, etc.)

dom4ha replied:
Sounds promising, but after a brief look it's not super clear to me how it works and whether it addresses all the problems.

The logic on the scheduler side may still be quite complex depending on which type of update is pending. For instance I suspect some of them may be blocking re-scheduling, but some not.

wojtek-t replied:
I think my comment could have been misleading.
It was supposed to be a meta-comment - i don't think we can really reuse any code from DeltaFIFO.
What I was trying to say is that it's effectively the same conceptual pattern that delta-fifo is using.

wojtek-t replied:
> The logic on the scheduler side may still be quite complex depending on which type of update is pending. For instance I suspect some of them may be blocking re-scheduling, but some not.

That's interesting - in particular:
1) what are the examples where an API call should prevent further attempts to do something with a given pod?
[The only one I can see is preemption, but it's not even for a given pod, so it doesn't belong to this category]

2) is there any case when if I have queued two different API calls for a given pod P - we actually want to send both, not just the last one?

sanposhiho replied:
> One example comes to my mind is preemption of just-bound pod. It would be safer to let the binding call happen first before we could start preemption process, to avoid a situation we preempt something not bound yet.

I guess it actually happens today? Because we're running the binding cycle asynchronously. 
Let's say pod-1 goes to the binding cycle (after "assume"), pod-2 triggers the preemption and then pod-2 could preempt pod-1 before pod-1's binding cycle completes.
And, what happens today (I guess) is, because pod-1 is deleted, the binding cycle for pod-1 would be just failing.

But, that is a good example showing the difficulty of this KEP: _Possibly_, not only API calls for the same object, but also API calls for different objects could be depending on one another...

macsko replied:
> 1. what are the examples where an API call should prevent further attempts to do something with a given pod?

I haven't found any such API call. Even preemption should not be a problem.

> 2. is there any case when if I have queued two different API calls for a given pod P - we actually want to send both, not just the last one?

We have 3 kinds of API calls for a Pod in scheduler: deletion (preemption), binding and status update (unschedulable or NNN). There is no such case where we would like to send two of these calls for a single Pod (see API calls categorization section).

alculquicondor commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/kep.yaml:
```diff
@@ -0,0 +1,28 @@
+  - dom4ha
+  - sanposhiho
+approvers:
+  - alculquicondor
```
Please change to @dom4ha or @sanposhiho

dom4ha commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,961 @@
+Making one universal approach of handling API calls in the kube-scheduler could allow these calls to be consistent, as well as better controlling
+the number of dispatched goroutines. Asynchronous preemption could also be migrated to this approach.
+
+### Goals
```
Looking from the high level, I think we may have the following goals (although it's not clear we're able to achieve all of them):

1. Make scheduling cycle free of blocking api-calls (any async option is fine)
2. Skip some type of updates if they soon become irrelevant by consecutive updates
3. Prioritize high importance updates (binding) over low importance ones if updates to the api-server gets throttled

Maybe we have more, but I think agreeing first on our goals is important before we can select a solution. One of my doubts here is that we may not know yet all the requirement we may have once we start designing workload-aware scheduling and reservations. More api calls may appear at that time.

wojtek-t replied:
+1 to it, but I think there is also a question about priorities. In my mental model:
1. eliminate blocking calls is P0
2. skip soon-to-be irrelevant updates is P1
3. prioritization of updates - is nice to have [I would like that but not necessary if that adds a lot of complexity]

sanposhiho replied:
I'm feeling like the third one is the next step, and we can just scope it out for now, at least from this KEP. Of course, we should keep it in mind when discussing this KEP though.
I agree that all of three are important, but it's just that we don't have to solve all of them at once on one KEP here; the third one looks complex enough to deserve being discussed on another KEP.

wojtek-t replied:
which seems to be roughly aligned with what I wrote above

dom4ha replied:
Agree, my with the above, the order was also in purpose.

dom4ha commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,961 @@
+
+Scheduling performance is crucial. One of the bottlenecks is the API calls done during the scheduling cycle. 
+The binding cycle is already asynchronous, but it would still be beneficial to re-evaluate whether the current model of busy-waiting goroutines is good long-term.
+The following operations involve pod-based API calls during scheduling:
```
It's great summary. It would be good to add a section that analyze and categorizes these calls against our goals. For instance which of them are part of the scheduling loop, which are high and which low importance that could be skipped.

macsko commented:
I added a section with API calls categorization and adjusted the goals. PTAL

To go further, we could probably agree on some high-level concept (queuing without blocking a pod?).

dom4ha commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,998 @@
+  if the newest status is stored in-memory.
+- API calls for non-Pod resources (5 - 10) should be further analyzed as they are not likely to consider the Pod-based API calls,
+  hence implementing those shouldn't block making (1 - 3) calls asynchronous.
+
```
> Is there any relevance order between calls for different pods?

I think there is because we need to protect the space freed up by preempted pods for the nominated pod. Sure, pods don't disapear immediately, but we shouldn't rely on it. That obviously puts in question whether we can priorities some updates over the others... which could change the order.

Alternatively we may still want to treat a preemption as one async operation that includes setting nomination and removing pods in the right sequence.

sanposhiho replied:
> because we need to protect the space freed up by preempted pods for the nominated pod. 

But, we store NNN to the nominator when we update NNN. i.e., even if we make the API call async, as long as we report NNN to the nominator synchronously, the space should be reserved, and not stealed by other pods' scheduling cycles.

wojtek-t replied:
Unless scheduler restarts in the meantime and we lose the in-memory state.

> Alternatively we may still want to treat a preemption as one async operation that includes setting nomination and removing pods in the right sequence.

which effectively introduces the ordering
The question is whether in-memory state is good enough, at least initially.

dom4ha commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,998 @@
+Both of the above API calls could be migrated to the new mechanism.
+
+In-tree plugins' operations that involve non-pod API calls during scheduling and could be made asynchronous,
+but not necessarily in the first place:
```
Moreover, we need to consider changes planned in the [Extended Resource KEP](https://github.com/kubernetes/enhancements/pull/5136), so it will be additional challenge how to put them together.

sanposhiho replied:
Never mind, I guess you meant ↓ described later in this kep.
> - Pod deletion caused by preemption (4) should cancel all Pod-based API calls for such a Pod.

dom4ha commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,998 @@
+There is no need to send two API calls for one Pod, because more relevant calls should override less relevant ones,
+and status updates can be combined into one call.
+
+In terms of API call priority, the order might be different (3rd goal):
```
```suggestion
In terms of API call priority, the order might be different (4th goal):
```

dom4ha commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,998 @@
+- Cannot be used for the `nominatedNodeName` scenario, requiring additional effort and separate handling.
+
+
+#### 1.3: Use advanced queue and don't block the pod from being scheduled in the meantime
```
I also agree that this should be the way to go. If we reach the consensus here, I'd move other options to the Alternatives considered and present one proposal.

sanposhiho commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,998 @@
+- Updating Pod status (1, 2) could be less important and called if there is space for it.
+  It's worth considering if setting `nominatedNodeName` (3) should have the same priority or higher,
+  because the higher delay might affect other components like Cluster Autoscaler.
+- API calls for non-Pod resources (5 - 10) could be analyzed case by case, but are likely less important than (5) and (4).
```
+1. I actually rather think API calls for non-Pod resources are likely equally important to 4 and 5 because, like described in 6-10 above, most of them are in the binding cycles, which are on the critical path, or PostFilter (similar to preemption - making space for pending pods).

sanposhiho replied:
But, the difficulty here is that we cannot just say "the API calls for non pods are likely important -> let's prioritize them" because obviously there'll be exceptions especially when it comes to the scheduler with custom plugins.
So, I basically agree with the KEP saying that this is case by case actually.

sanposhiho replied:
I guess it leads to another question: should we consider API calls being made within binding cycles? They're already async, so can we just treat them as a _single_ async group for multiple API calls? And, in this KEP, we only need to add a canceling functionality in the binding cycles.

sanposhiho commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,998 @@
+- Allows the pod to be scheduled again even before the API call completes.
+- Simplifies introducing new API calls to the kube-scheduler if the collision handling logic is configured correctly.
+
+Cons:
```
+ Memory consumption?

macsko replied:
I think all approaches will somehow result in increased memory consumption. For me, it's even hard to say which option will affect it the most.

sanposhiho commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,998 @@
+- Needs a clear strategy for how to update the in-memory pod object during scheduling.
+
+
+#### 2.3: Send API calls through a kube-scheduler's cache
```
I'm on this side. It's generally close to what we're doing today: assume() mechanism in the scheduling cache, or the nominator mechanism. 
Also, it's a lot easier to merge multiple updates into a single one. We always synchronously update the scheduler cache, and the cache asynchronously makes an API call under the hood. Even if there are two or more modifications made through the scheduler cache, the async worker can just make an API call with a final object state. And, the scheduler can work as if those modifications are immediately applied by referring to the scheduler cache (which, again, we're doing today.)

sanposhiho replied:
OTOH, what I don't like about the queue mechanism is, first, how to merge multiple updates into one. 
The KEP describes, regarding patch, they can apply the latest one. Really? Always? What if several modifications are conflicting? Also, if several fields are updated, should we compute a new patch diff every time?
Regarding binding, it says "Ignore status update API calls". Can we do that always for sure? For example, what if users want to add some conditions to the pods before binding phase? 

Something like that. So, I don't want to make a special rule to ignore/overwrite changes on the object, based on what we have at k/k.

Also, the second thing is what if users want to operate something at some point, and want to use it in another point? NominatedNodeName is a perfect example for this. The scheduler wants to put NNN so that the following scheduling cycles can refer to it. I know the scheduler, even today, has the nominator mechanism to cache it because the event handler doesn't give a NNN update immediately after making an API call. But, if we go with this cache option, we can generally solve that problem. 
Rather the queueing option would boost this problem because there could be multiple modifications on the fly, which are invisible to anyone until they're actually applied. 

---

I know I talked too many "what if". But, again though, given this mechanism is a standard general way to make API calls from the scheduler, we shouldn't put any assumption based on the in-tree plugins only.

wojtek-t replied:
OK - thinking more about that, I think you're right. I think the main argument is that we want to update scheduler cache with that information anyway. So my argument from https://github.com/kubernetes/enhancements/pull/5249#discussion_r2111217794 doesn't really work, because we need to touch the scheduler cache anyway.

macsko replied:
I'm now also thinking that cache idea is not really interchangeable with queue. Wouldn't we still need some kind of a queue behind a cache, as we don't want to send indefinite number of goroutines at one moment? We would also have to buffer some API calls in case a previous one for the same pod (or anything) is still running.

dom4ha replied:
Separation of nomination hints ownership (each component have its own hint) would help merging them and they would not conflict with scheduler changes.

Anyway, do we have other fields which can be modified by different components? For me it's one of the anti-patterns which obviously happens when NN was set manually, but it's also a known anti-pattern which users should avoid.

sanposhiho replied:
> we should allow the custom plugins to provide their own merging/conflicts resolution mechanism.

This is a good abstraction idea. 
Let's say it's called "merger"; we can have one merger per resource. We can implement default mergers for K8s resources that we're using at kube-scheduler, and we allow users to implement their own mergers for their resources, or if they need to update K8s default resources in a different manner, they can still implement custom mergers and disable default ones.

> But, if the failure is because of conflicting API calls, we won't be able to retry.

Yes, and also we cannot keep trying it indefinitely even for retryable errors. We need some proper way to handling those.

sanposhiho commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,998 @@
+- Aims for a consistent internal state of the pod within the kube-scheduler before calling the API, possibly simplifying conflict resolution.
+
+Cons:
+- Requires the cache to handle and merge updates coming from both the kube-scheduler's internal actions and external API events.
+- The cache currently only stores bound pods, requiring integration with the scheduling queue for pending pods.
+- Complex logic is needed to handle external updates arriving while an internal update is pending or in progress.
```
Regarding the first and the third cons: They're mentioned as cons here. However, this is a general problem, not a problem specific to this option since no option gracefully solves: how to handle the conflicts between changes the scheduler wants to make (but awaiting the scheduler to actually make an API call) vs the changes coming from outside. 

When such a conflict happens, which updates should be prioritized depends on which fields have to be updated on what purpose. That's a problem that I don't see an answer yet.

sanposhiho commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,998 @@
+- Simplifies introducing new API calls to the kube-scheduler if the collision handling logic is configured correctly.
+
+Cons:
+- Requires implementing complex, advanced queueing logic.
```
Really depends on what we're doing with this queueing logic. If it's just going to be the queue, obviously simple.
But, like mentioned at 2.2, if we are going to merge multiple operations into one, things will be getting complicated.

sanposhiho commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,998 @@
+- Needs a clear strategy for how to update the in-memory pod object during scheduling.
+
+
+#### 2.3: Send API calls through a kube-scheduler's cache
```
By the way, actually this cache idea looks like also part of `1: Where and how to handle API calls in the kube-scheduler` as well? 
Maybe it's not clear what to talk in 1 and what to talk in 2. Because, as we see, we're describing the queue thingy twice in both sections.

wojtek-t commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,998 @@
+- `nominatedNodeName` scenario support would require more effort in (1.1) or (1.2).
+
+
+#### 2.2: Make the API calls queued
```
Things may change behind the scene - so there might be races. But for a particular object relying on optimistic concurrency should just work.

If you're worries about scheduler preempting something and external actor changing NNN in the meantime - that has an inherent race in it no matter what we do - we may try to minimize the window, but it will always be there and I'm not sure I would actually try to optimize for it.

wojtek-t replied:
Update - after thinking about it I changed my mind and we may prefer 2.3 actually: https://github.com/kubernetes/enhancements/pull/5249#discussion_r2113342571

dom4ha commented:
> Also, regarding NNN part, can you please move your suggestions to NNN KEP PR, instead of here.

I did that already and there are more details there.

BenTheElder reviewed (commented):
This will need the PRR questionnaire at least the alpha requirements, PRR freeze is 
Thursday 12th June 2025 

https://www.kubernetes.dev/resources/release/

macsko commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,1003 @@
+
+Both of the above API calls could be migrated to the new mechanism.
+
+In-tree plugins' operations that involve non-pod API calls during scheduling and could be made asynchronous,
```
I changed it a bit now

macsko commented:
I added three proposals to design details section to be able to select the best queueing/caching approach (all are pretty similar). PTAL

I also already moved some previous proposals to Alternatives section.

BenTheElder commented on keps/prod-readiness/sig-scheduling/5229.yaml:
```diff
@@ -0,0 +1,3 @@
+kep-number: 5229
+beta:
+  approver: "@wojtek-t"
```
+1

dom4ha commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,886 @@
+
+This proposal could be implemented as a second step extension of proposal A.
+
+### Summary of API call management
```
I think it's worth to describe what happens on update failures, either due to conflicts or just reaching the max number or retries. For instance what happens when binding fails. Is it expected that a go routine waits for binding completion and reacts to a failure? Does the failed updates for which nobody waits, are ignored (like in status updates)?

macsko replied:
Added the examples to `Retrying API calls` section

dom4ha reviewed (commented):
Overall looks good. I think the proposal description is mainly missing a statement that some procedures will be still a separate go routines with ability to wait for the API calls to finish. This way we don't have to implement dependencies between those calls which depend on each other.

We should also make it clear which option we want to follow. IIUC option C is the one that is the most reasonable one and easy to implement.

macsko commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,886 @@
+
+- Prioritize high-importance updates (like binding) over low-importance ones if updates to the kube-apiserver get throttled.
+
+## Proposal
```
Added

macsko commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,886 @@
+e.g., don't set `nominatedNodeName` when Pod binding is enqueued.
+Initially, it could be a framework, which might be extended in the future, e.g., by introducing the possibility of setting delays.
+
+However, it is questionable what should happen if two update API calls for the same Pod are enqueued.
```
That's right, changed it to be more clear

wojtek-t commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,898 @@
+and if it's not expected, custom logic could always be added using an `APICall` interface.
+
+However, to support other potential use cases and have the newest object possible in the cache (proposals B and C, and optionally A),
+merging the object received by event handlers with API call details should also be added.
```
Note that merging with objects received from event handlers isn't a full solution. Given those are eventually consistent, there will always be a period of time, when the object was already changed and is not yet observed by the cache.

wojtek-t replied:
Agree - but propagating objects from handlers isn't the full solution - this may still happen given their eventually consistent nature.

wojtek-t replied:
Yes - we should do that.
But still that isn't a full solution - we can still face conflicts etc. that we need to handle.

macsko replied:
Yes, the scheduler could indeed operate on an older version of an object, but we face similar problems with the current implementation. The key is to ensure this is considered when handling an API call. However, it doesn't have any impact on current (Pod-based) API calls that we (will) made by the scheduler.

macsko replied:
> I think that we should have some rules which fields scheduler could blindly override and which requires pulling them in first and issue rescheduling of the conflicting changes. NodeName (assignment) is one of them

I don't feel we have much to do with that. For the actual use cases, overriding will be enough (maybe with handling the external binding/deletion separately). We could defer this discussion to the implementation phase and see what will be exactly needed then.

wojtek-t commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,898 @@
+It also should be defined how to handle such external updates if the API call is completed and the scheduler is waiting for the update to come in event handlers.
+The `ResourceVersion` of the object could be used to distinguish it, i.e., apply the API call details
+as long as the `ResourceVersion` of the received object is older than the version received by the update API call.
+Doing so would require storing the `ResourceVersion` of the updated object received from the API call somewhere in the cache or the queue.
```
nit: ResourceVersion is part of the object itself - why do you need to store it somehow additionally?

macsko replied:
That's true, I removed this part now.

wojtek-t commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,898 @@
+#### Enqueueing an API call while a previous one is in-flight
+
+One other possible scenario is when an API call is executing (is in-flight) and a new API call for the same object wants to be added.
+If both have the same call type, standard merging logic could be used, i.e., merge the new API call with the API call in flight.
```
I don't understand this - if the old one is already in-flight, how we can merge those?

macsko replied:
We need to keep the details of the API call in the map, even if the call is in-flight.

Let's say a status update API call is in-flight and wants to set the Pod condition `PodScheduled` to `false` with reason `foo`.
If then the next status update is queued for the same Pod, it can verify via merge if it actually changes anything. If it doesn't, it can be skipped. If it does, the merge should ignore changes that are duplicated with a call that is in-flight.
However, skip is a more important use case.

wojtek-t replied:
> If it does, the merge should ignore changes that are duplicated with a call that is in-flight.

This assumes that the request succeeds (or if not - will be retried). 

But anyway - I think I get your point - the wording is misleading though.
It's not really the we merge the new incoming request with the old one. It's that we adjust the new incoming one to reflect changes that are already in flight. 

It's a phrasing issue - merging for me means that we combine these two into a single one. Which is not what we're doing.

wojtek-t commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,898 @@
+###### Does enabling the feature change any default behavior?
+
+Pod scheduling might be retried even if the API call hasn't yet been executed.
+For instance, it might happen before the `PodScheduled` condition is set to `false`.
```
I don't understand this part - can you please clarify?

wojtek-t reviewed (commented):
Just some minor comments from me - other than that PRR lgtm.

dom4ha commented:
/approve

dom4ha commented:
/hold
/lgtm

sanposhiho reviewed (commented):
Left some comments, but LGTM overall. 
I have several unknowns around APIQueue, but I agree the overall direction and we can clarify such details at the implementation part.

macsko commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,915 @@
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
```
Good point, I added the section about it. I think we could partially support these plugins in the first version, but we have to carefully document allowed use cases.

wojtek-t commented:
/lgtm

Unholding based on https://github.com/kubernetes/enhancements/pull/5249#pullrequestreview-2937316668

/hold cancel

sanposhiho commented on keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md:
```diff
@@ -0,0 +1,915 @@
+
+// APICall describes the API call to be made and store all required data to make the call,
+// e.g. fields that should be updated or object to be added/removed.
+type APICall interface {
```
At the implementation, we need:
- how to prevent two different `APICall` are used for the same `APICallType`.
- how to replace a certain `APICall` that everyone is using. e.g., custom plugins may want to change how `status_update` is handled and hence the status_update's `APICall` implementation that is used by everyone.
  - meaning, no one should directly depend on a specific `APICall` implementation. Everyone should use each `APICall` implementation indirectly so that we can change it.

I _guess_ `APICall` implementations for each type should be stored properly somewhere accessible to everyone.
